src/topology/perfect.lean

```diff
@@ -0,0 +1,279 @@
+/-
+Copyright (c) 2022 Felix Weilacher. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Felix Weilacher
+-/
+import topology.separation
+import topology.bases
+
+/-!
+# Perfect Sets
+
+In this file we define perfect subsets of a topological space, and prove some basic properties,
+including a version of the Cantor-Bendixson Theorem.
+
+## Main Definitions
+
+* `acc_pt x C`: A point `x` is an accumulation point of a set `C`,
+  taken here to mean that every neighborhood of `x` contains points of `C` other than `x`
+* `perfect C`: A set `C` is perfect, meaning it is closed and every point of it
+  is an accumulation point of itself.
+
+## Main Statements
+
+* `perf_nonempty.splitting`: A perfect nonempty set contains two disjoint perfect nonempty subsets.
+  The main inductive step in the construction of an embedding from the Cantor space to a
+  perfect nonempty complete metric space.
+* `ctble_union_perfect_of_closed`: One version of the Cantor-Bendixson Theorem: A closed
+  set in a second countable space can be written as the union of a countable set and a perfect set.
+
+## Implementation Notes
+
+We do not require perfect sets to be nonempty, a condition which is often part of
+the definition of perfect. We include an extra definition `perf_nonempty`, which bundles
+these two conditions.
+
+We define a nonstandard predicate, `preperfect`, which drops the closed-ness requirement
+from the definition of perfect. In T1 spaces, this is equivalent to having a perfect closure,
+see `preperfect_iff_perfect_closure`.
+
+## References
+
+* [kechris1995] (Chapter 6)
+
+## Tags
+
+accumulation point, perfect set, Cantor-Bendixson.
+
+-/
+
+section
+
+open_locale topological_space filter
+open topological_space filter set
+
+variables {α : Type*} [topological_space α]
+
+/-- A point `x` is an accumulation point of a set `C` if every neighborhood
+of `x` contains a point of `C` other than `x`. See `acc_pt_iff`. -/
+def acc_pt (x : α) (C : set α) : Prop := ne_bot (𝓝[≠] x ⊓ 𝓟 C)
+
+/-- `x` is an accumulation point of `C` iff it is a cluster point of `C ∖ {x}`-/
+theorem acc_iff_cluster (x : α) (C : set α) : acc_pt x C ↔ cluster_pt x (𝓟({x}ᶜ ∩ C))
+  := by rw[acc_pt, nhds_within, inf_assoc, inf_principal, cluster_pt]
+
+/-- `x` is an accumulation point of `C` iff every neighborhood
+of `x` contains a point of `C` other than `x`. -/
+theorem acc_pt_iff_nhds (x : α) (C : set α) : acc_pt x C ↔ ∀ U ∈ 𝓝 x, ∃ y ∈ U ∩ C, y ≠ x
+  := by simp [acc_iff_cluster, cluster_pt_principal_iff, set.nonempty, exists_prop,
```

Comment:
In general we put the `:=` in the same line as the conclusion of the theorem, and line starting with `by` are not indented

Reply:
fixed